include/deal.II/matrix_free/fe_remote_evaluation.h

```diff
@@ -0,0 +1,768 @@
+   */
+  template <int dim, int n_components, typename value_type>
+  class PrecomputedFEEvaluationDataAccessor
+  {
+  public:
+    PrecomputedFEEvaluationDataAccessor(
+      const PrecomputedFEEvaluationData<dim, n_components, value_type> &data,
+      const PrecomputedFEEvaluationDataView                            &view)
+      : view(view)
+      , data(data)
+      , data_offset(numbers::invalid_unsigned_int)
+    {}
+
+    /**
+     * Get the value at quadrature point @p q. The entity on which the values
+     * are defined is set via `reinit()`.
+     */
+    const auto
+    get_value(const unsigned int q) const
+    {
+      Assert(data_offset != numbers::invalid_unsigned_int,
+             ExcMessage("reinit() not called."));
+      AssertIndexRange(data_offset + q, data.values.size());
+      return data.values[data_offset + q];
+    }
+
+    /**
+     * Get the gradients at quadrature point @p q. The entity on which the
+     * gradients are defined is set via `reinit()`.
+     */
+    const auto
+    get_gradient(const unsigned int q) const
+    {
+      Assert(data_offset != numbers::invalid_unsigned_int,
+             ExcMessage("reinit() not called."));
+      AssertIndexRange(data_offset + q, data.gradients.size());
+      return data.gradients[data_offset + q];
+    }
+
+
+    /**
+     * Set entity index at which quadrature points are accessed. This can, e.g.,
+     * a cell index, a cell batch index, or a face batch index.
+     */
+    void
+    reinit(const unsigned int index)
+    {
+      data_offset = view.get_shift(index);
+    }
+
+    /**
+     * Set cell and face_number at which quadrature points are accessed.
+     */
+    void
+    reinit(const unsigned int cell_index, const unsigned int face_number)
+    {
+      data_offset = view.get_shift(cell_index, face_number);
+    }
+
+  private:
+    /**
+     * PrecomputedFEEvaluationDataView provides information where values are
+     * located.
+     */
+    const PrecomputedFEEvaluationDataView &view;
+
+    /**
+     * PrecomputedFEEvaluationData stores the actual values.
+     */
+    const PrecomputedFEEvaluationData<dim, n_components, value_type> &data;
+
+    /**
+     * Offset to data after last call of `reinit()`.
+     */
+    unsigned int data_offset;
+  };
+
+
+} // namespace internal
+
+/**
+ * Communication objects know about the communication pattern.
+ * In case of (matrix-free) cells batches or faces batches
+ * a @c RemotePointEvaluation object stores the location of the
+ * remote points. @c batch_id_n_entities relates these points
+ * to the corresponding quadrature points of entity batches.
+ * For this the field stores batch IDs and the number of entities
+ * in the batch.
+ */
+template <int dim>
+struct FERemoteCommunicationObjectEntityBatches
+{
+  std::vector<std::pair<unsigned int, unsigned int>> batch_id_n_entities;
+  std::shared_ptr<Utilities::MPI::RemotePointEvaluation<dim>> rpe;
+
+  std::vector<std::pair<unsigned int, unsigned int>>
```

Comment:
Documentation is missing.

Reply:
[93ef439](https://github.com/dealii/dealii/pull/16300/commits/93ef43944b36a779518b004d86ae0a2d643d7ed2)